### ISSUE TITLE
Possible immutability issue?

### ISSUE THREAD

**Issue by andyjessop** [davidkpiano/xstate#58]
## Bug or feature request?
Possible bug

## Description:
I'm trying to create a little helper library to include actions, based off the example in the docs: http://davidkpiano.github.io/xstate/docs/#/guides/complete

```
export default (config: MachineConfig, actions?: StateXActions) => {
  const machine: StandardMachine|ParallelMachine = Machine(config);

  let currentState: StateValue = machine.initialState.value;

  const extendedState = {};

  const dispatch = (event: Event) => {
    const nextState = machine.transition(currentState, event, extendedState);

    nextState.actions
      .map(action => typeof action === 'string' ? action : action.type)
      .filter(key => actions && actions[key])
      .forEach(key => actions![key](event, dispatch));

    currentState = nextState.value;
  };

  return {
    currentState,
    dispatch,
    machine,
    state: extendedState,
  };
};
```

And trying to test with this:
```
test('dispatching go should enter go state', () => {
    const store = createStore(mocks.minimal);
    store.dispatch({ type: 'go' });
    expect(store.currentState).toEqual('go');
  });
```
mocks.minimal:
```
{
    initial: 'start',
    states: {
      start: {
        on: {
          go: 'go',
        },
      },
      go: {
        on: {
          stop: 'start',
        },
      },
    },
  },
```

## (Bug) Expected result:
I'm expecting that `store.currentState` will be updated with the next value.

## (Bug) Actual result:
The test fails:
```
Expected value to equal: "go"
Received: "start"
```
## (Bug) Potential fix:
I'm wondering if this is an issue with immutability. I've added the code to a repo here: https://github.com/andyjessop/statement if you want to reproduce.

**Comment by davidkpiano** [comment]
@andyjessop did you figure out the issue?

**Comment by andyjessop** [comment]
I didn't figure it out, but I figured it wasn't a problem with `xstate`. Sorry, I should have added it before closing.

At the point where I set `currentState = nextState.value` - that's all correct, so I guess it's something to do with how I'm passing it back from the closure. I thought it was because it was a primitive so was being passed by value, but then I changed it to pass the whole `State` instead of the string value, and I still got a similar result. 

So more debugging to do, but I don't think it's a problem with `xstate`.
